rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,386 @@
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_event_fini(&publisher_event);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_shutdown(this->context_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    ret = rcl_context_fini(this->context_ptr);
+    delete this->context_ptr;
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+protected:
+  rcl_context_t * context_ptr;
+  rcl_node_t * node_ptr;
+
+  rcl_publisher_t publisher;
+  rcl_event_t publisher_event;
+  rcl_subscription_t subscription;
+  rcl_event_t subscription_event;
+};
+
+bool
+wait_for_msgs_and_events(
+  rcl_subscription_t * subscription,
+  rcl_event_t * subscription_event,
+  rcl_event_t * publisher_event,
+  rcl_context_t * context,
+  size_t max_tries,
+  int64_t period_ms,
+  bool & msg_ready,
+  bool & subscription_event_ready,
+  bool & publisher_event_ready)
+{
+  msg_ready = false;
+  subscription_event_ready = false;
+  publisher_event_ready = false;
+
+  int num_subscriptions = (nullptr == subscription ? 0 : 1);
+  int num_events = (nullptr == subscription_event ? 0 : 1) + (nullptr == publisher_event ? 0 : 1);
+
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, num_subscriptions, 0, 0, 0, 0, num_events,
+    context, rcl_get_default_allocator());
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+    rcl_ret_t ret = rcl_wait_set_fini(&wait_set);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  size_t iteration = 0;
+  do {
+    ++iteration;
+    ret = rcl_wait_set_clear(&wait_set);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    if (nullptr != subscription) {
+      ret = rcl_wait_set_add_subscription(&wait_set, subscription, NULL);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    }
+    if (nullptr != subscription_event) {
+      ret = rcl_wait_set_add_event(&wait_set, subscription_event, NULL);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    }
+    if (nullptr != publisher_event) {
+      ret = rcl_wait_set_add_event(&wait_set, publisher_event, NULL);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    }
+
+    ret = rcl_wait(&wait_set, RCL_MS_TO_NS(period_ms));
+    if (ret == RCL_RET_TIMEOUT) {
+      continue;
+    }
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i] && wait_set.subscriptions[i] == subscription) {
+        msg_ready = true;
```

Comment:
Is it intentional that this function will wait the max number of iterations every time? It could return here or at the end of the do-while loop if not.

Reply:
Unintentional, good catch!